pkg/controller/seed-controller-manager/operating-system-profile-synchronizer/controller.go

```diff
@@ -0,0 +1,284 @@
+	if err := c.Watch(
+		&source.Kind{Type: &osmv1alpha1.OperatingSystemProfile{}},
+		&handler.EnqueueRequestForObject{},
+		kubermaticpred.ByNamespace(namespace),
+	); err != nil {
+		return fmt.Errorf("failed to create watch for operatingSystemProfiles: %w", err)
+	}
+
+	// Watch changes for OSPs and then enqueue all the clusters where OSM is enabled.
+	if err := c.Watch(
+		&source.Kind{Type: &kubermaticv1.Cluster{}},
+		enqueueOperatingSystemProfiles(reconciler.seedClient, reconciler.log, namespace),
+		workerlabel.Predicates(workerName),
+		withEventFilter(),
+	); err != nil {
+		return fmt.Errorf("failed to create watch for clusters: %w", err)
+	}
+	return nil
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("operatingsystemprofile", request.NamespacedName.String())
+	log.Debug("Reconciling")
+
+	osp := &osmv1alpha1.OperatingSystemProfile{}
+	if err := r.seedClient.Get(ctx, request.NamespacedName, osp); err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// OperatingSystemProfile is marked for deletion.
+	if osp.DeletionTimestamp != nil {
+		log.Debug("Deletion timestamp found for operatingSystemProfile")
+		if kuberneteshelper.HasFinalizer(osp, cleanupFinalizer) {
+			if err := r.handleDeletion(ctx, log, osp); err != nil {
+				err = fmt.Errorf("failed to delete operatingSystemProfile: %w", err)
+
+				log.Errorw("ReconcilingError", zap.Error(err))
+				r.recorder.Event(osp, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+
+				return reconcile.Result{}, err
+			}
+			return reconcile.Result{}, nil
+		}
+		// Finalizer doesn't exist so clean up is already done.
+		return reconcile.Result{}, nil
+	}
+
+	err := r.reconcile(ctx, log, osp)
+	if err != nil {
+		log.Errorw("Reconciling failed", zap.Error(err))
+		r.recorder.Event(osp, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+	}
+	return reconcile.Result{}, err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, osp *osmv1alpha1.OperatingSystemProfile) error {
+	if err := kuberneteshelper.TryAddFinalizer(ctx, r.seedClient, osp, cleanupFinalizer); err != nil {
+		return fmt.Errorf("failed to add finalizer: %w", err)
+	}
+	return r.syncAllUserClusterNamespaces(ctx, log, osp)
+}
+
+func (r *Reconciler) handleDeletion(ctx context.Context, log *zap.SugaredLogger, osp *osmv1alpha1.OperatingSystemProfile) error {
+	err := r.syncAllUserClusterNamespaces(ctx, log, osp)
+	if err != nil {
+		return err
+	}
+
+	// Remove the finalizer
+	return kuberneteshelper.TryRemoveFinalizer(ctx, r.seedClient, osp, cleanupFinalizer)
+}
+
+func (r *Reconciler) syncAllUserClusterNamespaces(ctx context.Context, log *zap.SugaredLogger, osp *osmv1alpha1.OperatingSystemProfile) error {
+	clusters := &kubermaticv1.ClusterList{}
+	if err := r.seedClient.List(context.Background(), clusters); err != nil {
```

Comment:
Please use the context you have available.

Reply:
Nice, let me update this.